pip/src/interpreter.rs

```diff
@@ -97,6 +101,31 @@ impl Interpreter {
         }
     }
 
+    /// Dumps the quantum state of the interpreter.
+    /// Returns a tuple of (amplitudes, num_qubits), where amplitudes is a dictionary from integer indices to
+    /// pairs of real and imaginary amplitudes.
+    fn dump_machine(&mut self, py: Python) -> (Py<PyDict>, usize) {
+        let (state, num_qubits) = self.interpreter.get_quantum_state();
```

Comment:
I think it'd be great if we went the extra mile to return a Python object here, maybe with a `.dict()` method on it to return the raw data as a dictionary (as you're doing here).

Because then we could implement the `__repr__` and `_repr_html_` methods to get bonus HTML formatting for notebooks. See: https://github.com/microsoft/qsharp/blob/34357178ece371eaa242bc7006130ec7b0b3f536/pip/src/interpreter.rs#L151-L175

Reply:
More great suggestions, thanks! I tried what you said and I think it worked well. It now returns a `StateDump` object that can be pretty printed, turned into a dict, or even indexed into. Here's an example:
![image](https://github.com/microsoft/qsharp/assets/10567287/8fb8da8c-baba-4d06-a1c0-e7df29e2d88e)